src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Helpers.cs

```diff
@@ -0,0 +1,872 @@
+                || (!BitConverter.IsLittleEndian && ((value - 0x0080u) & 0xFFFFu) < 0x0780u);
+        }
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff <paramref name="value"/> is a UTF-8 continuation byte;
+        /// i.e., has binary representation 10xxxxxx, where x is any bit.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUtf8ContinuationByte(in byte value)
+        {
+            // This API takes its input as a readonly ref so that the JIT can emit "cmp ModRM" statements
+            // directly rather than bounce a temporary through a register. That is, we want the JIT to be
+            // able to emit a single "cmp byte ptr [data], C0h" statement if we're querying a memory location
+            // to see if it's a continuation byte. Data that's already enregistered will go through the
+            // normal "cmp reg, C0h" code paths, perhaps with some extra unnecessary "movzx" instructions.
+            //
+            // The below check takes advantage of the two's complement representation of negative numbers.
+            // [ 0b1000_0000, 0b1011_1111 ] is [ -127 (sbyte.MinValue), -65 ]
+
+            return ((sbyte)value < -64);
+        }
+
+        /// <summary>
+        /// Given a 32-bit integer that represents two packed UTF-16 characters, all in machine-endian order,
+        /// returns true iff the two characters represent a well-formed UTF-16 surrogate pair.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsWellFormedUtf16SurrogatePair(uint value)
+        {
+            // Little-endian: Given [ LLLL HHHH ], validate that LLLL in [ DC00..DFFF ] and HHHH in [ D800..DBFF ].
+            // Big-endian: Given [ HHHH LLLL ], validate that HHHH in [ D800..DBFF ] and LLLL in [ DC00..DFFF ].
+            //
+            // We're essentially performing a range check on each component of the input in parallel. The allowed range
+            // ends up being "< 0x0400" after the beginning of the allowed range is subtracted from each element. We
+            // can't perform the equivalent of two CMPs in parallel, but we can take advantage of the fact that 0x0400
+            // is a whole power of 2, which means that a CMP is really just a glorified TEST operation. Two TESTs *can*
+            // be performed in parallel. The logic below then becomes 3 operations: "add/lea; test; jcc".
+
+            return (BitConverter.IsLittleEndian && ((value - 0xDC00_D800u) & 0xFC00_FC00u) == 0)
+                || (!BitConverter.IsLittleEndian && ((value - 0xD800_DC00u) & 0xFC00_FC00u) == 0);
+        }
+
+        /// <summary>
+        /// Converts a DWORD from machine-endian to little-endian.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint ToLittleEndian(uint value)
```

Comment:
Seems like a potential useful API for `BinaryPrimitives`?

Reply:
Tracked as https://github.com/dotnet/corefx/issues/36749.